cloverage/src/cloverage/instrument.clj

```diff
@@ -79,59 +93,79 @@
-    `#{defn}        :defn    ; don't expand defn to preserve stack traces
-    `#{defmulti}    :defmulti ; special case defmulti to avoid boilerplate
-    `#{defprotocol} :atomic   ; no code in protocols
-    `#{defrecord}   :record
-    `#{deftype*}    :deftype*
-    `#{ns}          :atomic
-
-    ;; http://dev.clojure.org/jira/browse/CLJ-1330 means AOT-compiled definlines
-    ;; are broken when used indirectly. Work around - do not wrap the definline
-    `#{booleans bytes chars shorts floats ints doubles longs} :inlined
-    atomic-special?   :atomic
-    ;; XXX: we used to not do anything with unknown specials, now we wrap them
-    ;; in a macro, then macroexpand back to original form. Methinks it's ok.
-    special-symbol?   :unknown
-    :list))
-
-(defn list-type-in-env [[head & _] env]
-  ;; if the list is a call to a symbol with a local binding, but not a special form such as `var` or `new` (which
-  ;; aren't overshadowed by local variables) return `:list` type.
-  (if (and (get env head)
-           (not (special-symbol? head)))
-    :list
-    (list-type head)))
+(def ^:private list-head-symbol->type
+  "Map of symbols to the type of form they represent when they're at the head of a list or list-like form.
+
+    {'do :do, 'let* :let, 'try :try, ...}"
+  (into {} (for [[symbols typ]
+                 ;; namespace-less symbols
+                 {'#{try}         :try  ; try has to special case catch/finally
+                  '#{if do throw} :do   ; these special forms can recurse on all args
+                  '#{let* loop*}  :let
+                  '#{def}         :def  ; def can recurse on initialization expr
+                  '#{fn*}         :fn
+                  '#{set!}        :set  ; set must not evaluate the target expr
+                  '#{.}           :dotjava
+                  '#{case*}       :case*
+                  '#{new}         :new
+                  '#{reify*}      :reify*
+                  ;; FIXME: monitor-enter monitor-exit
+
+                  ;; namespaced macros
+                  `#{cond}        :cond ; special case cond to avoid false partial
+                  `#{loop let}    :let
+                  `#{letfn}       :letfn
+                  `#{for doseq}   :for
+                  `#{fn}          :fn
+                  `#{defn}        :defn     ; don't expand defn to preserve stack traces
+                  `#{defmulti}    :defmulti ; special case defmulti to avoid boilerplate
+                  `#{defprotocol} :atomic   ; no code in protocols
+                  `#{defrecord}   :record
+                  `#{deftype*}    :deftype*
+                  `#{ns}          :atomic
+
+                  ;; http://dev.clojure.org/jira/browse/CLJ-1330 means AOT-compiled definlines
+                  ;; are broken when used indirectly. Work around - do not wrap the definline
+                  `#{booleans bytes chars shorts floats ints doubles longs} :definlined-fn-call}
+                 symb symbols]
```

Comment:
The fix in this PR mainly concerns functions defined with `defn` but with `:inline` metadata such as `+` or `int`. Cloverage previously had a special case for functions defined with `definline` such as `booleans` or `bytes`. For these existing special cases, I did not change the behavior, but to make the difference between the two situations clearer I renamed the `:inline` form type to `:defininlined-fn-call` and added `:inlined-fn-call` for the forms fixed in this PR.

Reply:
I appreciate this comment, it made it much easier to follow what was going on.